clang/include/clang/AST/Type.h

```diff
@@ -486,15 +486,12 @@ class Qualifiers {
   /// Returns true if the address space in these qualifiers is equal to or
   /// a superset of the address space in the argument qualifiers.
   bool isAddressSpaceSupersetOf(Qualifiers other) const {
-
-    return
-        isAddressSpaceSupersetOf(getAddressSpace(), other.getAddressSpace()) ||
-        (!hasAddressSpace() &&
-         (other.getAddressSpace() == LangAS::sycl_private ||
-          other.getAddressSpace() == LangAS::sycl_local ||
-          other.getAddressSpace() == LangAS::sycl_global ||
-          other.getAddressSpace() == LangAS::sycl_constant ||
-          other.getAddressSpace() == LangAS::sycl_generic));
+    return isAddressSpaceSupersetOf(getAddressSpace(),
+                                    other.getAddressSpace()) ||
+           (!hasAddressSpace() &&
+            (other.getAddressSpace() == LangAS::opencl_private ||
```

Comment:
This seems strange. Why do you need this downstream change here?

Reply:
What are you trying to achieve?

Reply:
I'd like to avoid adding AS qualifiers to all types and treat all objects of "unqualified" types to be allocated in "generic" address space. This allows us to emit valid SPIR-V files for the SYCL code, which inter-operates regular C++ code satisfying SYCL kernel restrictions.
We qualify with address spaces only types required by the SPIR-V spec (e.g. kernel parameters, program scope variables, etc). All other objects (not qualified explicitly by user via SYCL pointer classes), are residing in "generic" address space. This approach allows us altering C++ type system.

Reply:
In OpenCL (and other languages too) we do qualifiers deduction during parsing. All pointers without explicit address space are in generic in OpenCL v2.0. I guess you could just do the same.

This function determines qualifier relations regardless of exact type. Also this function should just have the same behaviour as the one above since they are to be used interchangeably. I think we just added this function to allow calling it as a member but it just called the helper. So I don't understand why these two have different logic in your downstream changes. They should return the same result for the same pair of qualifiers.

Reply:
We have recently changed how deduction works:
https://reviews.llvm.org/D65744

My guess it you just alter `deduceOpenCLPointeeAddrSpace` slightly and make sure you invoke it in the same places.

Reply:
> In OpenCL (and other languages too) we do qualifiers deduction during parsing. All pointers without explicit address space are in generic in OpenCL v2.0. I guess you could just do the same.

I have an alternative proposal. I think we should remove `opencl_generic` semantic address space and treat "default" address space (aka "no address space"/"unqualified type") as "generic" in Sema for OpenCL/SYCL mode. This will allows us to fix this particular issue + remove OpenCL custom logic from the C/C++ compiler, w/o losing the quality of the compiler diagnostics or generated code (in case of SYCL it will be even an improvement you suggesting).

In addition to that we should disallow using `generic` and `__generic` keywords. These are not defined by the OpenCL spec and reserved for future use, but I see these are exposed in the language.

Reply:
Also note: Our justification for doing what @bader says is that setting a default address space that exists in the type system absolutely wrecks havok on template metaprograms in C++ mode.  Programmers do all sorts of weird things to get to their types, that it becomes really difficult to properly cover all the cases with setting a 'default'.

Reply:
> > In OpenCL (and other languages too) we do qualifiers deduction during parsing. All pointers without explicit address space are in generic in OpenCL v2.0. I guess you could just do the same.
> 
> I have an alternative proposal. I think we should remove `opencl_generic` semantic address space and treat "default" address space (aka "no address space"/"unqualified type") as "generic" in Sema for OpenCL/SYCL mode. This will allows us to fix this particular issue + remove OpenCL custom logic from the C/C++ compiler, w/o losing the quality of the compiler diagnostics or generated code (in case of SYCL it will be even an improvement you suggesting).
> 

Perhaps just to clarify whether you mean removing generic or default address space here? For generic it is doable but since they are only used for pointers/references I am not sure what advantage this brings. For default address space I don't see an easy viable path. The address space has to be created somewhere and if it's not a job of the frontend then something else will have to do this anyway. I don't think it is an easy task no matter at which stage in the compiler it is done and considering that the generality of compilation i.e. providing flow without optimisations or with linking of separate modules it feels that the earlier we add the address space the better.

Can you explain the problems that you encounter with generic address space?

> In addition to that we should disallow using `generic` and `__generic` keywords. These are not defined by the OpenCL spec and reserved for future use, but I see these are exposed in the language.

I think the keywords will have to stay for backwards compatibility. We can always map it to something different in AST or IR. That's no problem.

Reply:
> Also note: Our justification for doing what @bader says is that setting a default address space that exists in the type system absolutely wrecks havok on template metaprograms in C++ mode. Programmers do all sorts of weird things to get to their types, that it becomes really difficult to properly cover all the cases with setting a 'default'.

In Clang we are currently working on providing concrete and well defined semantic to address spaces in C++. It is not just for OpenCL but for memory segmented architectures in general. It is a big task but not impossible. This will allow to provide a language feature that developers can understand and use reliably and that implementers can support in a portable way.

Reply:
I'd be interested to see where that goes.  Clang ends up creating pointer/reference types in a large number of places and people do a TON of metaprogramming with it, so something like std::is_same ends up getting broken really easily when we deduced address-spaces.

Reply:
For that we need to add "remove address space" metaprogramming utility that developers should understand to use in memory segmented architectures or perhaps even library implementer can just add in their implementations underneath. Here is related discussion about t:
https://reviews.llvm.org/D69938#1747262

However, I struggled to find time to add it by now although I think it should be pretty straight forward. We do have a prototype from OpenCL C++ that might be possible to reuse:
https://github.com/KhronosGroup/libclcxx/blob/master/include/__ocl_type_traits_address_spaces.h

Is there something similar in SYCL?

Reply:
For SYCL we need to support most existing C++ code including a lot of the standard libraries.  So new metaprogramming utilities are insufficient unfortunately.

Reply:
Is this even feasible? Do you mean on host or device?

Accelerators are extremely constrained so you can't just compile any C++ code that standard libraries would use for device side. That is the reason why standards like OpenCL, CUDA etc are defined to constrain features that can't be used.

The host side should not be affected much by address spaces?

Reply:
For both. In device mode, we were putting the address space, so non-emitted (not used in kernel) code with some template metaprogramming was causing a bunch of issues.  It wasn't something we could hack the compiler around, because it was often type-comparisons via template deduction.

Additionally, SYCL supports as much existing C++ as we can, so when it comes to metaprogramming we try to support everything since there's no harm in it.  So saying we don't support some non-address-space aware template code that doesn't cause anything 'bad' to be emitted seemed like the wrong direction.

Reply:
I am not sure this is a tractable problem to solve. If you allow any C++ library to be compiled for device than you should just allow to use any C++ to compile on the device. There is no need for any dialect like OpenCL or SYCL.

>  It wasn't something we could hack the compiler around, because it was often type-comparisons via template deduction.

Right. I don't see the point in hacking the compiler to work around conceptual problems. It would fall over somewhere anyway.

Templates for example are purely language/syntactic features so I assume language/frontend should provide some definition/solution to how it works?

> So saying we don't support some non-address-space aware template code that doesn't cause anything 'bad' to be emitted seemed like the wrong direction.

This presumably will compile the code that would generate IR with no address spaces? That wouldn't work on some accelerators that have memory segments. Potentially you should define the range of targets that SYCL supports. But it doesn't seem to me at this point if you just omit address spaces it would be ok for any OpenCL device.

Reply:
We are able to 'compile' just about all code in the frontend, so just about any library is supported unless it does a handful of illegal things.  We DO generate IR with no address space, but have a pass (IIRC) that goes through and adds 'generic' to all otherwise unlabeled pointers.  I believe we had an idea to do this in Clang CodeGen, but I don't know what ever happened with that.

So, essentially, we treat the "no" address space as "generic" all the time, but we do it later in the process in order to not mess with SEMA/type building.

Reply:
> We are able to 'compile' just about all code in the frontend, so just about any library is supported unless it does a handful of illegal things. We DO generate IR with no address space, but have a pass (IIRC) that goes through and adds 'generic' to all otherwise unlabeled pointers. I believe we had an idea to do this in Clang CodeGen, but I don't know what ever happened with that.
> 
> So, essentially, we treat the "no" address space as "generic" all the time, but we do it later in the process in order to not mess with SEMA/type building.

Do you do this just for the C++ standard libraries then I guess all pointers should be 'unlabeled'? However things might get interesting if you combine with  application code that calls the libs. I guess you need to make sure to insert the conversions to generic? However it might not be legal for some cases i.e. constant address space. Do you provide support for it?

In the application code however I presume such address space inference might be dangerous. What if the pointer was qualified by private address space that is the same as default?

Reply:
No, we do it for all 'unlabeled' pointers.  I believe we have code that just treats 'generic' as AS-0 (the unspecified version) so they end up being the same, that way all comparisons are the same.  

I don't see how it is any different from the way you suggested of making unspecified==generic, just less work.  We still enforce conversions the same way (with some modifications because the rules weren't perfectly the same), we just don't differentiate between generic and 'none'.

Reply:
> I don't see how it is any different from the way you suggested of making unspecified==generic, just less work. We still enforce conversions the same way (with some modifications because the rules weren't perfectly the same), we just don't differentiate between generic and 'none'.

As a matter of fact I didn't suggest this at all. Mapping to generic is VERY different to me from turning something that might not have been generic to generic. You can end up making generic something that had the address space explicitly i.e. private. Although I don't entirely know the details of what you are doing. I will leave it with you.

In general I don't see it often when software that isn't designed for a particular architecture is going to just recompile and work for a different one. In fact libraries are often performance optimized to a particular platform so not sure how performant your solution would be.

Reply:
> We DO generate IR with no address space, but have a pass (IIRC) that goes through and adds 'generic' to all otherwise unlabeled pointers. I believe we had an idea to do this in Clang CodeGen, but I don't know what ever happened with that.

I don't think this is accurate. We map default address space in AST to addrspace(4) in IR (generic) in Clang CG.

Reply:
Do you bypass the address space deduction in Clang completely? I imagine you would then end up with stack allocations from addrspace(4)?

Reply:
> Do you bypass the address space deduction in Clang completely? 

Yes. 

> I imagine you would then end up with stack allocations from addrspace(4)?

AFAIK, alloca address space in LLVM IR is determined by a data layout. They end up in addrspace(0) in our case.

The overall idea is basically:
  1) All pointers in AST have the same address space (default). This way we follow standard C++ type system, so every valid C++ code can be compiled and executed.
  2) In Clang CG, concrete address space pointers appear: alloca produces private pointers, kernel arguments can be global or local. Since AST assumes that all pointers have the same type, most of these concrete pointers are immediately casted to generic.
  3) Later in the compilation flow we use llvm::InferAddressSpaces pass to optimize addrspace casts  where the algorithm can identify the real address space behind a generic pointer.

One notable exception to (1) is SYCL multi_ptr class: it is a wrapper class around a *qualified* pointer (similar to shared_ptr and uniq_ptr). Since this is a class and not a raw pointer, it follows the rules of C++, and all operations around his qualified pointer happen in a "controlled environment" (i.e. no template metamagic can interfere).

Reply:
> All pointers in AST have the same address space (default).

I presume if pointer address space was provided explicitly you would keep it?

> This way we follow standard C++ type system, so every valid C++ code can be compiled and executed.

What about C++ features that can't be easily supported by OpenCL devices like virtual functions?

> In Clang CG, concrete address space pointers appear: alloca produces private pointers, kernel arguments can be global or local. Since AST assumes that all pointers have the same type, most of these concrete pointers are immediately casted to generic.

I don't see how they appear in CodeGen if you didn't have them in AST at all? I presume you have some address spaces somewhere in AST? Also how about address spaces that can't convert to generic like `constant`. Do you support this?

> Later in the compilation flow we use llvm::InferAddressSpaces pass to optimize addrspace casts where the algorithm can identify the real address space behind a generic pointer.

That makes sense. I guess the limitation is if you compile separate translation modules or don't inline function calls then the inference will be limited.

Reply:
> > All pointers in AST have the same address space (default).
> 
> I presume if pointer address space was provided explicitly you would keep it?

Yes, but the only way to do that is through sycl::multi_ptr (and sycl::accessor, forgot about it) classes. User never writes 

> 
> > This way we follow standard C++ type system, so every valid C++ code can be compiled and executed.
> 
> What about C++ features that can't be easily supported by OpenCL devices like virtual functions?

We do support function pointers on CPU and GPU as an extension:
https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SPIRV/SPV_INTEL_function_pointers.asciidoc

> 
> > In Clang CG, concrete address space pointers appear: alloca produces private pointers, kernel arguments can be global or local. Since AST assumes that all pointers have the same type, most of these concrete pointers are immediately casted to generic.
> 
> I don't see how they appear in CodeGen if you didn't have them in AST at all? I presume you have some address spaces somewhere in AST? 

Sorry, I should've say this better.
You're correct, some address spaces are present in AST, but they are "hidden" from a user inside C++ classes, so there is no legal way to get their type.
Alloca as I mentioned, is always addrspace(0) from data layout, this address space does not come from AST.

> Also how about address spaces that can't convert to generic like `constant`. Do you support this?

Constant is not fully supported as a raw pointer. It cannot be converted to generic, so if we allow it, then we risk emitting LLVM IR that cannot be lowered to SPIR-V.
Constant can still be used with sycl::multi_ptr and sycl::accessor classes though.

> 
> > Later in the compilation flow we use llvm::InferAddressSpaces pass to optimize addrspace casts where the algorithm can identify the real address space behind a generic pointer.
> 
> That makes sense. I guess the limitation is if you compile separate translation modules or don't inline function calls then the inference will be limited.

Right. More limitations come if a pointer is stored in memory and then loaded elsewhere.
Unless mem2reg eliminates this store/load, it is difficult to prove anything.

Reply:
> We do support function pointers on CPU and GPU as an extension:
> https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SPIRV/SPV_INTEL_function_pointers.asciidoc

Right this is great! But I was thinking of generic supported by multiple vendors. Does it mean that SYCL will define a new type of devices that it support i.e. not just any OpenCL device?

> Sorry, I should've say this better.
> You're correct, some address spaces are present in AST, but they are "hidden" from a user inside C++ classes, so there is no legal way to get their type.
> Alloca as I mentioned, is always addrspace(0) from data layout, this address space does not come from AST.

I am thinking of some language related aspect that you might miss to diagnose with this strategy. Let's say if user code creates pointers:
loc_ptr1 - pointer to local addr space
loc_ptr2 - pointer to local addr space
glob_ptr - pointer to global addr space

then are then cast to generic (no address space) straight away and then if they appear in the following statement:

loc_ptr1 = somevariable ? loc_ptr2 : glob_ptr;

if you don't have address spaces any more at this point you can't really provide an error saying that the code is likely illegal?

Are you aware of the cases like that and how severe they might be?

Reply:
> > We do support function pointers on CPU and GPU as an extension:
> > https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SPIRV/SPV_INTEL_function_pointers.asciidoc
> 
> Right this is great! But I was thinking of generic supported by multiple vendors. Does it mean that SYCL will define a new type of devices that it supports i.e. not just any OpenCL device?

Since OpenCL doesn't have this in the core specification, it is hard (impossible?) to provide a solution that works everywhere (with reasonable performance). But the extension is fairly generic, so I guess any vendor interested in this feature can implement it.

> 
> > Sorry, I should've say this better.
> > You're correct, some address spaces are present in AST, but they are "hidden" from a user inside C++ classes, so there is no legal way to get their type.
> > Alloca as I mentioned, is always addrspace(0) from data layout, this address space does not come from AST.
> 
> I am thinking of some language related aspect that you might miss to diagnose with this strategy. Let's say if user code creates pointers:
> 
> loc_ptr1 - pointer to local addr space
> loc_ptr2 - pointer to local addr space
> glob_ptr - pointer to global addr space

Address space qualifiers are not exposed in the SYCL language, so there is no way for a user to create a qualified raw pointer. (unless they use internal attributes directly; nobody should do this)

> 
> then are they cast to generic (no address space) straight away and then if they appear in the following statement:
> 
> loc_ptr1 = somevariable ? loc_ptr2 : glob_ptr;
> 
> if you don't have address spaces any more at this point you can't really provide an error saying that the code is likely illegal?

This is perfectly legal, if you write it without address space qualifiers.
Consider this (not real SYCL code, but close enough):
```c++
void main() {
  global_ptr<int> glob_ptr;
  local_ptr<int> loc_ptr;
  int cond = random();
  parallel_for([]() {
    // kernel code
    int *glob = glob_ptr.get();  // glob_ptr.get returns an int* in default AS (== generic)
    int *loc= loc_ptr.get();     // loc_ptr.get also returns an int* in default AS
    int *ptr = cond? glob : loc; // ptr is also default
    *ptr = 42; // this is going to be a store by a generic pointer, unless the compiler
               // figures out what `cond' is.
  }
}
```

In LLVM IR this looks like:
```
define kernel(i32 addrspace(1)* glob_ptr, i32 addrspace(3)* loc_ptr, i32 cond) {
    %glob = addrspacecast %glob_ptr to i32 addrspace(4)*
    %loc = addrspacecast %loc_ptr to i32 addrspace(4)*
    %ptr = select %cond,  i32 addrspace(4)* %glob,  i32 addrspace(4)* %loc
    store  i32 addrspace(4)* %ptr, 42
}
```

Again, this a valid code for a device that supports generic address space. If a device doesn't support it, then compilation terminates (or gives a diagnostic) from the device backend/middle-end compiler (it is not as pretty as the a diagnostic from Clang). Although, for less capable devices, it is probably better to use C++ classes instead of raw pointers where possible.

> 
> Are you aware of the cases like that and how severe they might be?

Control flow, stored/loaded pointers from memory, arrays of pointers in different address spaces. I suspect these cases are pretty common for generic C++ code, and there is a runtime overhead for supporting them. User can optimize them manually (by using sycl::multi_ptr class), or rely on the compiler.

Reply:
In  my original code I didn't use generic pointer. By adding generic you removed the ability to diagnose the behaviour that is illegal i.e. assigning local address space pointer into global.

Even in OpenCL 2.0 compatible devices address spaces are used for performance reasons so developers might want to avoid generic as much as possible. Then it is really helpful to get diagnostics from the compiler as debugging such issues can be very painful to nearly impossible on accelerators.

Ok I generally understand the reasoning behind your design. I just find it a bit unfortunate that you remove address spaces from AST early so you can't benefit from the power of semantical analysis fully. I don't know whether developers will find it valuable that the compiler compiles everything but then they get little help in what goes wrong.